pkg/frontend/querymiddleware/retry.go

```diff
@@ -87,10 +92,33 @@ func (r retry) Do(ctx context.Context, req Request) (Response, error) {
 			lastErr = err
 			log := util_log.WithContext(ctx, spanlogger.FromContext(ctx, r.log))
 			level.Error(log).Log("msg", "error processing request", "try", tries, "err", err)
+
+			if errors.Is(err, util.NotRunningError{}) && tries < r.maxRetries-1 {
+				// We're still starting up. Wait a little until retrying.
+				time.Sleep(r.notRunningBackoff)
+			}
+
 			continue
 		}
 
 		return nil, err
 	}
 	return nil, lastErr
 }
+
+func (r retry) isTerminalError(err error) bool {
+	if apierror.IsNonRetryableAPIError(err) {
+		return true
+	}
+
+	if errors.Is(err, context.Canceled) {
+		return true
+	}
+
+	notRunningError := util.NotRunningError{}
+	if errors.As(err, &notRunningError) {
+		return notRunningError.State == services.Stopping || notRunningError.State == services.Terminated || notRunningError.State == services.Failed
```

Comment:
Would it not be better to check if that state is `starting` or `new` to return False? Otherwise we consider any other service error to be terminal? (in case more states are added in the future for example).

Reply:
Good point, I've changed this.